### PR TITLE
Adds pre-purchase approval to ERC20 Locks

### PR DESCRIPTION
**Pull request by akeem**
# Description

* Moves getLock & getAddress up in the object hierarchy, supporting downstream usecases.
* Purchase key now requests approval prior to the purchase when an address has been provided

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by akeem** [comment]
> Do you mind not changing `getLock` and `getAddressBalance` since this is not needed for now?

@julien51 this has been addressed, please review when you have a moment

File: unlock-js/src/erc20.js [hunk 1]
```diff
@@ -16,6 +16,25 @@ export async function getErc20BalanceForAddress(
   return utils.toNumber(balance)
 }
 
+export async function approveTransfer(
```
Comment (julien51) [inline unlock-js/src/erc20.js]: Do you mind adding tests for this?

File: unlock-js/src/v10/purchaseKey.js [hunk 2]
```diff
@@ -15,8 +16,19 @@ import Errors from '../errors'
...
+export default async function(
+  lockAddress,
+  owner,
+  keyPrice,
+  account,
+  data,
+  erc20Address
+) {
   const lockContract = await this.getLockContract(lockAddress)
+  if (erc20Address) {
```
Comment (julien51) [inline unlock-js/src/v10/purchaseKey.js]: Can we add a test for that?

File: unlock-js/src/__tests__/erc20.test.js [hunk 3]
```diff
@@ -42,4 +42,11 @@ describe('erc20', () => {
       expect(balance).toEqual(1337)
     })
   })
+
+  describe('approveTransfer', () => {
+    // this functionality will need to be tested after we revist how we test
+    //this functionality. Ideally we should test the contents of the transaction
+    // that will be sent.
+    it.skip('should dispatch the appropriate transaction', async () => {})
```
Comment (julien51) [inline unlock-js/src/__tests__/erc20.test.js]: We need this one ;)

Comment (akeem) [reply unlock-js/src/__tests__/erc20.test.js]: @julien51 In the current state of the way the tests are structured I don't believe it is possible to create this in a way that actually provides confidence here. 

We can mock the network requests. mock the mining of a block, and finally, then return. Ideally being able to assert the content of the transaction or its hash should be sufficient here.  Let's discuss this in person

File: unlock-js/src/__tests__/erc20.test.js [hunk 4]
```diff
@@ -67,4 +72,64 @@ describe('erc20', () => {
...
+
+  describe('approveTransfer', () => {
+    it('should dispatch the appropriate transaction', async () => {
+      expect.assertions(2)
+
+      let gas = '0x73b3'
+      let approvalAmount = 2
+      let transactionHash =
+        '0xbb1f660b4a40e1931b1206d15421f364487b494920b2e67152a9ff4c74dcb66d'
+      let transactionData =
```
Comment (julien51) [inline unlock-js/src/__tests__/erc20.test.js]: Ideally you would actually decompose this so we know what's in there?

**Pull request merged**